### ISSUE TITLE
[Bug] Cannot start container after adding SOCI index

### ISSUE THREAD

**Issue by mohamed-haidara-cko** [awslabs/soci-snapshotter#829]
### Description

I'm starting an ECS service with an EFS filesystem on Fargate. The entire project can be found here: https://github.com/haidaraM/terraform-jenkins-aws-fargate/pull/7. You can run it if you have Terraform installed and an AWS account.

<details>
<summary>Dockerfile</summary>

```dockerfile
FROM --platform=linux/amd64 jenkins/jenkins:2.420

USER root

# Need awscli to get jenkins configuration from S3
RUN apt update && curl "https://awscli.amazonaws.com/awscli-exe-linux-x86_64.zip" -o "awscliv2.zip" && \
    unzip awscliv2.zip && ./aws/install && rm -rf awscliv2.zip ./aws && aws --version

USER jenkins

COPY plugins.txt /usr/share/jenkins/ref/plugins.txt
RUN jenkins-plugin-cli --verbose --plugin-file /usr/share/jenkins/ref/plugins.txt

COPY ./controller-entrypoint.sh /usr/bin/entrypoint

ENTRYPOINT ["/usr/bin/entrypoint"]
```

</details>

<details>
<summary>controller-entrypoint.sh</summary>

```shell
#!/bin/bash

set -e
set -x

JENKINS_CONF_LOCAL_PATH="${JENKINS_HOME}/jenkins.yaml"

if [[ -n "${JENKINS_CONF_S3_URL}"  ]]; then
  echo "Getting Jenkins configuration from ${JENKINS_CONF_S3_URL} for JCAS plugin"
  # this file will be read by jenkins configuration as code plugin at EVERY START UP of the Controller
  aws s3 cp "${JENKINS_CONF_S3_URL}" "${JENKINS_CONF_LOCAL_PATH}"
else
  # see https://github.com/jenkinsci/configuration-as-code-plugin/issues/825
  echo "Removing ${JENKINS_CONF_LOCAL_PATH} to avoid overriding existing configuration"
  rm -f "${JENKINS_CONF_LOCAL_PATH}"
fi

# entrypoint from the docker image
exec tini -- /usr/local/bin/jenkins.sh # <------ Container blocks here!

```

</details>

After generating a SOCI index for my image, ECS cannot start it. It is blocked on the command `exec tini -- /usr/local/bin/jenkins.sh` (see the controller-entrypoint.sh above).
I used the [Containerized SOCI Index Builder](https://github.com/aws-samples/aws-fargate-seekable-oci-toolbox/blob/main/containerized-index-builder/README.md) on a M1 Macbook. 

**However**, If I first start the service from [the image on Docker Hub](https://hub.docker.com/layers/elmhaidara/jenkins-aws-fargate/2.420/images/sha256-75dc09b0a90a4c5f9899a71d43c3559526841e39cf55ce26f006b0803c7db06e?context=explore) (empty EFS), and re-redeploy the service with the image on ECR with the index, it works (non-empty EFS).

### Describe the results you expected

The container should start even though the EFS is empty. 

### Host information

1. OS: MacOS M1 Ventura 13.5.1
2. Snapshotter Version: soci version v0.3.0 2574f6466b4a148be04f8f79c6e71167994452a7
3. Containerd Version: ctr github.com/containerd/containerd v1.7.3

**Comment by Kern--** [comment]
To clarify a few details:

1) Does the container hang forever or is it just much slower to start?
2) The setup is that you're trying to run Jenkins with the $JENKINS_HOME directory on EFS, right? It works if you've pre-populated EFS, right? Is there any chance you can run it without EFS? Does that still cause the container to hang forever?

Do you happen to have a more minimal reproduction?

**Comment by Kern--** [comment]
Actually, I was able to reproduce this using your Dockerfile, entrypoint script, and the plugin list from your repo with a local snapshotter.

I modified the Docker to install strace and ran it with:
```
sudo nerdctl run --rm -it --net host --snapshotter soci --entrypoint strace --cap-add SYS_PTRACE $IMAGE -f -o /tmp/trace /usr/bin/entrypoint
```

From the trace, I can see it getting stuck in an `lseek` loop:
```
29    lseek(3, 1774757, SEEK_SET)       = 1774757
29    lseek(3, 1774757, SEEK_DATA)      = 1774757
29    lseek(3, 1774757, SEEK_HOLE)      = 1774757
29    lseek(3, 1774757, SEEK_SET)       = 1774757
29    lseek(3, 1774757, SEEK_DATA)      = 1774757
29    lseek(3, 1774757, SEEK_HOLE)      = 1774757
29    lseek(3, 1774757, SEEK_SET)       = 1774757
29    lseek(3, 1774757, SEEK_DATA)      = 1774757
29    lseek(3, 1774757, SEEK_HOLE)      = 1774757
29    lseek(3, 1774757, SEEK_SET)       = 1774757
```

This is a known issue in one of our dependencies: https://github.com/hanwen/go-fuse/issues/487. We have a PR out to fix that bug, but this is blocked until that gets merged.

**Comment by Kern--** [comment]
Looking a bit deeper at the trace, SOCI was treating the file as a sparse (hence the SEEK_DATA/SEEK_HOLE flags), but overlayFS wasn't. Something seems off about that, so I ran `stat` on the file both with an without SOCI:

OverlayFS:
```
$ stat /usr/share/jenkins/ref/plugins/apache-httpcomponents-client-4-api.jpi
  File: /usr/share/jenkins/ref/plugins/apache-httpcomponents-client-4-api.jpi
  Size: 1774757   	Blocks: 3472       IO Block: 4096   regular file
```

SOCI:
```
$ stat /usr/share/jenkins/ref/plugins/apache-httpcomponents-client-4-api.jpi
  File: /usr/share/jenkins/ref/plugins/apache-httpcomponents-client-4-api.jpi
  Size: 1774757     Blocks: 434        IO Block: 4096   regular file
```

OverlayFS is correctly reporting the number of 512-byte blocks, but SOCI is reporting the number of `IO Block`-byte blocks.

https://github.com/awslabs/soci-snapshotter/blob/main/fs/layer/node.go#L817

That's a bug in SOCI that we can fix. Even without that fix, there shouldn't be any functional difference between SOCI and OverlayFS (ignoring the current upstream bug), but this is probably causing a bunch of unnecessary lstats looking for data in sequential files.

**Comment by mohamed-haidara-cko** [comment]
Hi Kern,

Thanks for considering this issue. In case it still helps, here is the answer to your question:

1. The container hangs forever and is killed by ECS. 
2. The mount point in the container is `JENKINS_HOME=/var/jenkins_home`. 
 
> It works if you've pre-populated EFS, right? 

Yes, it works. Basically, if I run the service with an image on DockerHub and switch to ECR with SOCI, it works. It doesn't work If I first run with ECR + SOCI and empty EFS.

> Is there any chance you can run it without EFS? Does that still cause the container to hang forever?

I can try it if still needed.

**Comment by mohamed-haidara-cko** [comment]
Hi @Kern-- 

Sorry to comment on a closed issue (I can open a new one if needed) but I'm still having the same issue with the updated version. I built [this commit](https://github.com/awslabs/soci-snapshotter/commit/342eff60b4eee144fe7b92a125a79a301e53725a) locally using [this Dockerfile](https://github.com/haidaraM/terraform-jenkins-aws-fargate/blob/fargate-soci/docker/containerized-index-builder/Dockerfile) (the image is here on [Dockerhub](https://hub.docker.com/layers/elmhaidara/soci-index-builder/342eff60b4eee144fe7b92a125a79a301e53725a/images/sha256-3879881c1c47541fedbe5c8b9b15f339a93c5c025eedf5bced3c0d1dd39b012a?context=repo)), pushed the index and started the service: the container is not able to start. 

I also tested the service without the EFS same behavior.
